content/blogs/gui-setup-part3-first-testing-problems.md

```diff
@@ -0,0 +1,267 @@
+_However_, Wine code also uses a temporary destination file to perform the copy, instead of directly copying over any possible existing file. See below for the reason why.)
+
+The default `TEMP` directory for the currently-running ReactOS installation is retrieved by a call to `GetTempPathW()` -- then followed by an attempt to open the source with `LZOpenFileW()` that also serves to handle single-compressed source files.
+(Remember those `MSPAINT.EX_`-like files in your Windows &le; 2003 installation disks? Those are compressed single-file CAB archives, containing one single file, in this example: `MSPAINT.EXE`.)
+This call will succeed even if the file is _not_ compressed.
+
+Then, a temporary destination file path is generated with `GetTempFileNameW()`, based on the retrieved `TEMP` directory path.
+Finally the source file is extracted to this temporary file (or just copied to, in case it was not compressed).
+Only then (omitted code that follows the snippet above), the created temporary file is moved to its final target.
+(The reason why this temporary file step is necessary, is for handling copy/replacements over files that could be already existing and currently in use.)
+
+The question is then: _**Where is the default**_ `TEMP` directory, when running the ReactOS setup via the ReactOS LiveCD?
+The answer is: _**Located** in the currently-running ReactOS instance, that is, **in the read-only LiveCD ISO itself!!**_ (in `X:\reactos\TEMP`).
+Obviously, attempting to create/write a file to this destination will inevitably fail.
+(Even though the ultimate destination of the extracted & copied file will be on the write-accessible ReactOS partition on the hard-disk.)
+
+Because of this failure, attempting to install such files will fail.
+This failure, coupled to the fact that the [custom setupapi file-copy queue callback function](https://learn.microsoft.com/en-us/windows/win32/setupapi/creating-a-custom-queue-callback-routine) I wrote in the GUI installer, did not correctly handle for file-copy failures (and instead was just requesting setupapi to re-attempt copying the same file), led to an infinite loop and thus an apparent hanging.
+
+#### The fix
+
+Fixing this bug is actually pretty simple.
+
+First, the [custom setupapi file-copy queue callback function](https://learn.microsoft.com/en-us/windows/win32/api/setupapi/nc-setupapi-psp_file_callback_w) in the GUI installer needs to recognize it is called due to a copy error
+(with the [`SPFILENOTIFY_COPYERROR`](https://learn.microsoft.com/en-us/windows/win32/setupapi/spfilenotify-copyerror) setupapi [notification](https://learn.microsoft.com/en-us/windows/win32/setupapi/notifications))
+and return `FILEOP_SKIP` or any other "error" value, instead of the default `FILEOP_DOIT` or `FILEOP_RETRY`.
+This would make the code more "future-proof" for any other sort of copy errors. (An improvement would also to show a custom error message box.)
```

Comment:
```suggestion
This would make the code more "future-proof" for any other sort of copy errors. (An improvement would be also to show a custom error message box.)
```
(also, one sentence per line - the one in braces)